azure-iot-device/azure/iot/device/product_info.py

```diff
@@ -0,0 +1,50 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import platform
+from azure.iot.device.constant import VERSION, IOTHUB_IDENTIFIER, PROVISIONING_IDENTIFIER
+
+python_runtime = platform.python_version()
+os_type = platform.system()
+os_release = platform.version()
+architecture = platform.machine()
+
+
+class ProductInfo(object):
+    """
+    A class for creating product identifiers or agent strings for IotHub as well as Provisioning.
+    """
+
+    @staticmethod
+    def _get_common_user_agent():
+        return "({python_runtime};{os_type} {os_release};{architecture})".format(
+            python_runtime=python_runtime,
+            os_type=os_type,
+            os_release=os_release,
+            architecture=architecture,
+        )
+
+    @staticmethod
+    def get_iothub_user_agent():
+        """
+        Create the user agent for IotHub
+        """
+        return (
+            "{iothub_iden}/{version}".format(iothub_iden=IOTHUB_IDENTIFIER, version=VERSION)
+            + ProductInfo._get_common_user_agent()
```

Comment:
Instead of using format and then concatenating strings, how about 
`
        return (
            "{iothub_iden}/{version}{common}".format(
            iothub_iden=IOTHUB_IDENTIFIER, 
            version=VERSION, 
            common=ProductInfo._get_common_user_agent(),
        ))
`

Reply:
I'm mentioning this because the way you have it now, there's a heap allocation that immediately gets freed (the result of the format call).  It's not a big deal on cpython, but it can lead to heap fragmentation.  

---
In reply to: [384152287](https://github.com/Azure/azure-iot-sdk-python/pull/464#discussion_r384152287) [](ancestors = 384152287)